price-feeder/oracle/provider/okx.go

```diff
@@ -0,0 +1,210 @@
+	OkxTickerPair struct {
+		InstId string `json:"instId"` // Instrument ID ex.: BTC-USDT
+		Last   string `json:"last"`   // Last traded price ex.: 43508.9
+		Vol24h string `json:"vol24h"` // 24h trading volume ex.: 11159.87127845
+	}
+
+	// OkxTickerResponse defines the response structure of a Okx ticker
+	// request.
+	OkxTickerResponse struct {
+		Data []OkxTickerPair   `json:"data"`
+		Arg  SubscriptionTopic `json:"arg"`
+	}
+
+	// SubscriptionTopic Topic with the ticker to be subscribed/unsubscribed
+	SubscriptionTopic struct {
+		Channel string `json:"channel"` // Channel name ex.: tickers
+		InstId  string `json:"instId"`  // Instrument ID ex.: BTC-USDT
+	}
+
+	// SubscriptionMsg Message to subscribe/unsubscribe with N Topics
+	SubscriptionMsg struct {
+		Op   string              `json:"op"` // Operation ex.: subscribe
+		Args []SubscriptionTopic `json:"args"`
+	}
+)
+
+// NewOkxProvider creates a new OkxProvider
+func NewOkxProvider(ctx context.Context, logger zerolog.Logger, pairs ...types.CurrencyPair) (*OkxProvider, error) {
+	wsURL := url.URL{
+		Scheme: "wss",
+		Host:   okxHost,
+		Path:   okxPath,
+	}
+
+	wsConn, _, err := websocket.DefaultDialer.Dial(wsURL.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to Okx websocket: %w", err)
+	}
+
+	provider := &OkxProvider{
+		wsURL:            wsURL,
+		wsClient:         wsConn,
+		logger:           logger.With().Str("module", "oracle").Logger(),
+		tickers:          map[string]OkxTickerPair{},
+		msReadNewMessage: msReadNewMessage,
+	}
+
+	if err := provider.subscribeTickers(pairs...); err != nil {
+		return nil, err
+	}
+
+	go provider.handleReceivedTickers(ctx)
+
+	return provider, nil
+}
+
+// GetTickerPrices returns the tickerPrices based on the saved map
+func (p *OkxProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]TickerPrice, error) {
+	tickerPrices := make(map[string]TickerPrice, len(pairs))
+
+	for _, currencyPair := range pairs {
+		price, err := p.getTickerPrice(currencyPair)
+		if err != nil {
+			return nil, err
+		}
+
+		tickerPrices[currencyPair.String()] = price
+	}
+
+	return tickerPrices, nil
+}
+
+func (p *OkxProvider) getTickerPrice(cp types.CurrencyPair) (TickerPrice, error) {
+	instrumentId := getInstrumentId(cp)
+	tickerPair, ok := p.tickers[instrumentId]
+	if !ok {
+		return TickerPrice{}, fmt.Errorf("failed to get %s", instrumentId)
+	}
+
+	return tickerPair.toTickerPrice()
+}
+
+func (p *OkxProvider) handleReceivedTickers(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Duration(p.msReadNewMessage) * time.Millisecond):
+			// time after to avoid asking for prices too frequently
+			messageType, bz, err := p.wsClient.ReadMessage()
+			if err != nil {
+				// if some error occurs continue to try to read the next message
+				p.logger.Debug().Msg(fmt.Sprintf("Okx provider could not read message error: %+v - bytes: %s", err, bz))
```

Comment:
@RafilxTenfen

nit: We use a structured logger (zerolog) which is powerful and efficient because it has typed fields. Using `Sprintf` defeats the purpose here.

Instead do: `Err(err).Msg("Okx provider could not read message")

Reply:
Thanks for explaining, done at https://github.com/umee-network/umee/pull/522/commits/a891ff6bfeffd73711e69fb368c22a9064805a91